### PR TITLE
Throw error when an argument of incorrect `BIT_SIZE` is passed to `maskl` and `maskr`

### LINKED ISSUE #4081: Throw error when I in intrinsic `maskr` /`maskl` is greater than BIT_SIZE of INTEGER(KIND=4)
```fortran
program intrinsic_150
    print*, maskr(63)
    print*, maskr(54)
end
```
```console
$ gfortran a.f90 && ./a.out
a.f90:2:18:

    2 |     print*, maskr(63)
      |                  1
Error: ‘I’ at (1) must be less than or equal to the BIT_SIZE of INTEGER(KIND=4)
a.f90:3:18:

    3 |     print*, maskr(54)
      |                  1
Error: ‘I’ at (1) must be less than or equal to the BIT_SIZE of INTEGER(KIND=4)
$ lfortran a.f90
2147483647
4194303
```

### PR DESCRIPTION
**Pull request by kmr-srbh**
fixes #4081

We throw an error:
1. When the `BIT_SIZE` of the passed argument is incorrect.
2. When the passed argument is negative.

### Incorrect `BIT_SIZE`
```fortran
program example
   print*, maskl(63)
end
```
```console
(lf) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
semantic error: `I` must be less than or equal to the BIT_SIZE of INTEGER(KIND=4)
 --> ./examples/example.f90:2:12
  |
2 |    print*, maskl(63)
  |            ^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

### Negative argument
```fortran
program example
   print*, maskr(-24)
end
```
```console
(lf) saurabh-kumar@Awadh:~/Projects/System/lfortran$ lfortran ./examples/example.f90
semantic error: `I` must be nonnegative
 --> ./examples/example.f90:2:12
  |
2 |    print*, maskr(-24)
  |            ^^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

The changes are identical for both `maskl` and `maskr`.

### CODE REVIEW

**Comment by assem2002** [comment]
You can push reference tests. I think it would be perfect and ready to go then.

File: src/libasr/pass/intrinsic_functions.h [hunk 1]
```diff
@@ -2782,11 +2782,16 @@ namespace Popcnt {
...
     static ASR::expr_t* eval_Maskl(Allocator& al, const Location& loc,
-            ASR::ttype_t* t1, Vec<ASR::expr_t*>& args, diag::Diagnostics& /*diag*/) {
+            ASR::ttype_t* t1, Vec<ASR::expr_t*>& args, diag::Diagnostics& diag) {
         int32_t kind = ASRUtils::extract_kind_from_ttype_t(t1);
         int64_t i = ASR::down_cast<ASR::IntegerConstant_t>(args[0])->m_n;
-        if (((kind == 4) && i > 32) || (kind == 8 && i > 64) || i < 0) {
-                return nullptr;
+        if (((kind == 4) && i > 32) || (kind == 8 && i > 64)) {
+            diag.semantic_error_label("`I` must be less than or equal to the BIT_SIZE of INTEGER(KIND=" 
+            + std::to_string(kind) + ")", {loc}, "");
```
Comment (certik) [inline src/libasr/pass/intrinsic_functions.h:2790]: Let's use lowercase names in error messages.

Comment (certik) [reply src/libasr/pass/intrinsic_functions.h:2790]: Also what is `I`? I assume that is the first argument? If so, I would mention this in the error message.

**Comment by certik** [comment]
We also need to add the same checks for the runtime implementation.

**Review by Shaikh-Ubaid — APPROVED** [review]
Ondrej's suggestions need to be incorporated. Apart from that this looks good to me. Thanks for this!

**Comment by HarshitaKalani** [comment]
@kmr-srbh I think it is easy to complete now. Can you please work on it and get it done?

File: src/libasr/pass/intrinsic_functions.h (hunk 1 above)
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_functions.h:2790]:
```suggestion
        if ((kind == 4 && i > 32) || (kind == 8 && i > 64)) {
            diag.semantic_error_label(" first argument of `maskl` must be less than or equal to the BIT_SIZE of INTEGER(KIND=" 
            + std::to_string(kind) + ")", {loc}, "");
```

File: src/libasr/pass/intrinsic_functions.h [hunk 2]
```diff
@@ -2782,11 +2782,16 @@ namespace Popcnt {
...
+            return nullptr;
+        } else if (i < 0) {
+            diag.semantic_error_label("`I` must be nonnegative", {loc}, "");
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_functions.h:2793]:
```suggestion
            diag.semantic_error_label("first argument of `maskl` must be nonnegative", {loc}, "");
```

File: src/libasr/pass/intrinsic_functions.h [hunk 3]
```diff
@@ -2821,11 +2826,16 @@ namespace Maskl {
...
+            return nullptr;
```
Comment (HarshitaKalani) [inline src/libasr/pass/intrinsic_functions.h:2838]: Similarly here.

File: tests/errors/maskl_incorrect_bit_size.f90 [hunk 4]
```diff
@@ -0,0 +1,3 @@
+program example
+   print*, maskl(63)
+end
```
Comment (HarshitaKalani) [inline tests/errors/maskl_incorrect_bit_size.f90:3]:
```suggestion
program maskl_incorrect_bit_size
   print*, maskl(63)
end program
```

File: src/libasr/pass/intrinsic_functions.h (hunk 3 above)
Comment (kmr-srbh) [reply src/libasr/pass/intrinsic_functions.h:2838]: Done.

**Comment by kmr-srbh** [comment]
I do not know how to handle the runtime implementation for this. I see that a function call is being constructed in the ASR for this intrinsic. How do we handle this in the back-end?

File: tests/errors/maskl_negative.f90 [hunk 5]
```diff
@@ -0,0 +1,3 @@
+program maskl_negative
+   print*, maskl(-24)
+end
```
Comment (HarshitaKalani) [inline tests/errors/maskl_negative.f90:3]:
```suggestion
end program
```

File: tests/errors/maskr_incorrect_bit_size.f90 [hunk 6]
```diff
@@ -0,0 +1,3 @@
+program maskr_incorrect_bit_size
+   print*, maskr(63)
+end
```
Comment (HarshitaKalani) [inline tests/errors/maskr_incorrect_bit_size.f90:3]:
```suggestion
end program
```

File: tests/errors/maskr_negative.f90 [hunk 7]
```diff
@@ -0,0 +1,3 @@
+program maskr_negative
+   print*, maskr(-24)
+end
```
Comment (HarshitaKalani) [inline tests/errors/maskr_negative.f90:3]:
```suggestion
end program
```

**Review by HarshitaKalani — APPROVED** [review]
You don't have to do that for runtime. The PR looks good to me, just minor changes need to be handled.

**Comment by kmr-srbh** [comment]
> You don't have to do that for runtime. The PR looks good to me, just minor changes need to be handled. And rebase the PR over main. Thanks for this.

@HarshitaKalani this implementation would not work for runtime values. Could you please have a look at https://github.com/lfortran/lfortran/pull/4084#issuecomment-2130659736?

**Comment by kmr-srbh** [comment]
I have a plan for a runtime implementation. I will test it out and see how it lands. This will probably help us in throwing error messages for all the intrinsics for runtime values.

**Comment by HarshitaKalani** [comment]
```fortran
program main
   integer :: x = 63
   integer :: y = -24
   print*, maskl(x)
   print*, maskl(y)
end program
```
```console
$ lfortran a.f90
-2
-16777216
$ gfortran a.f90 && ./a.out
          -2
   -16777216
```
The code works for runtime values with GFortran too.
I asked chatgpt for the reason and here's what I got:

In case of compile time evaluation, the argument 63 is a constant and is evaluated at compile time. The maskl function expects its argument to be within the range of valid bit positions for its integer kind. For a default 32-bit integer (INTEGER(KIND=4)), the valid range is 0 to 31 because BIT_SIZE of a 32-bit integer is 32. Since 63 is outside this range, gfortran throws an error at compile time:
In case of runtime evaluation, although 63 is still outside the valid range for the bit position, gfortran handles this by effectively taking the modulo of the argument with 32 (the BIT_SIZE of the integer). In this case, 63 % 32 = 31, and maskl(31) is evaluated, which returns -2 for a 32-bit integer.

**Summary**
**Compile-time constant arguments**: The Fortran compiler checks the validity of the argument against the BIT_SIZE and throws an error if it's out of range.
**Runtime variable arguments**: The value is evaluated at runtime, and out-of-range values are handled by taking the modulo with BIT_SIZE, thus avoiding a runtime error.

**Review by certik — APPROVED** [review]
I think that's fine.

**Comment by HarshitaKalani** [comment]
@kmr-srbh can you rebase it and resolve the conflicts and it'll be ready to merge.

**Pull request merged**